File: impl/csharp/Hex.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Text;

namespace Naalp
{
    /// <summary>Lowercase, no-separator hex encode/decode (the wire byte-field encoding).</summary>
    public static class Hex
    {
        private const string Digits = "0123456789abcdef";

        /// <summary>Encode bytes to lowercase hex.</summary>
        public static string Encode(byte[] data)
        {
            var sb = new StringBuilder(data.Length * 2);
            foreach (byte b in data)
            {
                sb.Append(Digits[(b >> 4) & 0xF]);
                sb.Append(Digits[b & 0xF]);
            }
            return sb.ToString();
        }

        /// <summary>Decode a hex string (any case) to bytes.</summary>
        public static byte[] Decode(string s)
        {
            if (s == null)
            {
                throw new NaalpException("Malformed", "hex string is null");
            }
            if ((s.Length & 1) != 0)
            {
                throw new NaalpException("Malformed", "hex string has odd length");
            }
            byte[] outp = new byte[s.Length / 2];
            for (int i = 0; i < outp.Length; i++)
            {
                outp[i] = (byte)((Nibble(s[2 * i]) << 4) | Nibble(s[2 * i + 1]));
            }
            return outp;
        }

        private static int Nibble(char c)
        {
            if (c >= '0' && c <= '9') return c - '0';
            if (c >= 'a' && c <= 'f') return c - 'a' + 10;
            if (c >= 'A' && c <= 'F') return c - 'A' + 10;
            throw new NaalpException("Malformed", "invalid hex character '" + c + "'");
        }
    }
}

File: impl/java/src/main/java/sh/bubblefish/naalp/Hex.java
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp;

/** Lowercase-hex encode/decode for the byte-valued wire fields of the conformance protocol. */
public final class Hex {
    private static final char[] DIGITS = "0123456789abcdef".toCharArray();

    private Hex() {}

    public static String encode(byte[] b) {
        char[] out = new char[b.length * 2];
        for (int i = 0; i < b.length; i++) {
            int v = b[i] & 0xFF;
            out[2 * i] = DIGITS[v >>> 4];
            out[2 * i + 1] = DIGITS[v & 0x0F];
        }
        return new String(out);
    }

    public static byte[] decode(String s) {
        if ((s.length() & 1) != 0) {
            throw new NaalpException("Malformed", "odd-length hex string");
        }
        int n = s.length() / 2;
        byte[] b = new byte[n];
        for (int i = 0; i < n; i++) {
            int hi = Character.digit(s.charAt(2 * i), 16);
            int lo = Character.digit(s.charAt(2 * i + 1), 16);
            if (hi < 0 || lo < 0) {
                throw new NaalpException("Malformed", "invalid hex digit");
            }
            b[i] = (byte) ((hi << 4) | lo);
        }
        return b;
    }
}
